src/Common/tests/System/Net/VirtualNetwork/VirtualNetwork.cs

```diff
@@ -36,7 +36,24 @@ public void ReadFrame(bool server, out byte[] buffer)
 
             semaphore.Wait(WaitForReadDataTimeoutMilliseconds);
 
-            bool dequeueSucceeded = packetQueue.TryDequeue(out buffer);
+            bool dequeueSucceeded = false;
+            int remainingRetries = 3;
+            int backOffDelayMilliseconds = 10;
+
+            do
+            {
+                dequeueSucceeded = packetQueue.TryDequeue(out buffer);
```

Comment:
The assert that got removed was the first thing that failed under stress after the test got switched to proper async/await.

The wait is on a semaphore, not a lock. It's acceptable to have multiple threads adding while multiple other threads are reading.

TryDequeue is, by contract, allowed to fail: http://stackoverflow.com/questions/5645348/under-what-conditions-can-trydequeue-and-similar-system-collections-concurrent-c

With this fix, the second TryDequeue always seems to pass. (Given the issue above I would have seen tests hanging for >27hrs which didn't happen overnight.)